### PR TITLE
Bug and test fixes

### LINKED ISSUE #1314: NRE in HandleModuleImportStar
Using the setup in #364, hovering over `__path__` gives:

```
Opening document file:///c:/Users/jabaile/test_projects/issue364/app.py
Analysis of app(User) queued
Import:  parent.child.one c:\Users\jabaile\test_projects\issue364\parent\child\one.py 
Analysis version 2 of 1 entries has started.
Analysis of parent.child.one(User) queued
Analysis of app(User) failed. Exception message: Object reference not set to an instance of an object..
Analysis version 2 of 1 entries has been completed in 17.2121 ms.
Analysis version 4 of 2 entries has started.
Analysis of parent.child.one(User) on depth 0 completed for library in 8.1233 ms.
Analysis of app(User) failed. Exception message: Object reference not set to an instance of an object..
Analysis complete: 2 modules in 67.6934 ms.
Analysis version 4 of 2 entries has been completed in 15.2549 ms.
complete.
Hover in file:///c:/Users/jabaile/test_projects/issue364/app.py at (2, 15)
[Error - 1:00:14 PM] Request textDocument/hover failed.
  Message: Object reference not set to an instance of an object.
  Code: -32000 
   at Microsoft.Python.Analysis.Analyzer.Handlers.ImportHandler.HandleModuleImportStar(PythonVariableModule variableModule) in C:\Users\jabaile\python-language-server\src\Analysis\Ast\Impl\Analyzer\Handlers\FromImportHandler.cs:line 79
   at Microsoft.Python.Analysis.Analyzer.Handlers.ImportHandler.AssignVariables(FromImportStatement node, IImportSearchResult imports, PythonVariableModule variableModule) in C:\Users\jabaile\python-language-server\src\Analysis\Ast\Impl\Analyzer\Handlers\FromImportHandler.cs:line 62
   at Microsoft.Python.Analysis.Analyzer.Handlers.ImportHandler.HandleFromImport(FromImportStatement node) in C:\Users\jabaile\python-language-server\src\Analysis\Ast\Impl\Analyzer\Handlers\FromImportHandler.cs:line 45
   at Microsoft.Python.Parsing.Ast.FromImportStatement.Walk(PythonWalker walker) in C:\Users\jabaile\python-language-server\src\Parsing\Impl\Ast\FromImportStatement.cs:line 56
   at Microsoft.Python.Parsing.Ast.SuiteStatement.Walk(PythonWalker walker) in C:\Users\jabaile\python-language-server\src\Parsing\Impl\Ast\SuiteStatement.cs:line 41
   at Microsoft.Python.Parsing.Ast.PythonAst.Walk(PythonWalker walker) in C:\Users\jabaile\python-language-server\src\Parsing\Impl\Ast\PythonAst.cs:line 81
   at Microsoft.Python.Analysis.Analyzer.PythonAnalyzerSession.AnalyzeEntry(IDependencyChainNode`1 node, PythonAnalyzerEntry entry, IPythonModule module, PythonAst ast, Int32 version) in C:\Users\jabaile\python-language-server\src\Analysis\Ast\Impl\Analyzer\PythonAnalyzerSession.cs:line 322
   at Microsoft.Python.Analysis.Analyzer.PythonAnalyzerSession.Analyze(IDependencyChainNode`1 node, AsyncCountdownEvent ace, Stopwatch stopWatch) in C:\Users\jabaile\python-language-server\src\Analysis\Ast\Impl\Analyzer\PythonAnalyzerSession.cs:line 256
--- End of stack trace from previous location where exception was thrown ---
   at Microsoft.Python.Analysis.Analyzer.PythonAnalyzerEntry.<>c.<GetAnalysisAsync>b__25_0(Task`1 t) in C:\Users\jabaile\python-language-server\src\Analysis\Ast\Impl\Analyzer\PythonAnalyzerEntry.cs:line 99
   at System.Threading.Tasks.ContinuationResultTaskFromResultTask`2.InnerInvoke()
   at System.Threading.ExecutionContext.RunInternal(ExecutionContext executionContext, ContextCallback callback, Object state)
--- End of stack trace from previous location where exception was thrown ---
   at System.Threading.Tasks.Task.ExecuteWithThreadLocal(Task& currentTaskSlot)
--- End of stack trace from previous location where exception was thrown ---
   at Microsoft.Python.Analysis.Analyzer.PythonAnalyzer.GetAnalysisAsync(PythonAnalyzerEntry entry, CancellationToken timeoutCt, CancellationToken cancellationToken) in C:\Users\jabaile\python-language-server\src\Analysis\Ast\Impl\Analyzer\PythonAnalyzer.cs:line 98
   at Microsoft.Python.Analysis.Analyzer.PythonAnalyzer.GetAnalysisAsync(IPythonModule module, Int32 waitTime, CancellationToken cancellationToken) in C:\Users\jabaile\python-language-server\src\Analysis\Ast\Impl\Analyzer\PythonAnalyzer.cs:line 91
   at Microsoft.Python.LanguageServer.Documents.Document.GetAnalysisAsync(Uri uri, IServiceContainer services, Int32 msTimeout, CancellationToken cancellationToken) in C:\Users\jabaile\python-language-server\src\LanguageServer\Impl\Documents\Document.cs:line 35
   at Microsoft.Python.LanguageServer.Implementation.Server.Hover(TextDocumentPositionParams params, CancellationToken cancellationToken) in C:\Users\jabaile\python-language-server\src\LanguageServer\Impl\Implementation\Server.Editor.cs:line 56
   at Microsoft.Python.LanguageServer.Implementation.LanguageServer.Hover(JToken token, CancellationToken cancellationToken) in C:\Users\jabaile\python-language-server\src\LanguageServer\Impl\LanguageServer.cs:line 221
```

### LINKED ISSUE #1318: PathUtils.IsNormalizedPath and PathUtils.NormalizePath work incorrectly with UNC
`PathUtils.IsNormalizedPath` always returns false.
`PathUtils.NormalizePath` throws `IndexOutOfRangeException`

### LINKED ISSUE #1320: Incorrect relative import handling in Python 2x
For the `package` with `__init__.py` in lib and user code:
```
├─ folder
│   ├─ package.py
│   └─ two.py
├─ package.py
└─ one.py
```
`import package` in `one` should import `package` from lib, while in `two` it should import `package.py` from the `folder`.

### LINKED ISSUE #1321: If one user paths is part of the name of another, PathResolverSnapshot is built incorrectly
It works if paths are `c:\github` and `c:\github\src`, but fails if they are `c:\github\src` and `c:\github\src2`

### PR DESCRIPTION
**Pull request by AlexanderSher**
- Fix #1314: NRE in HandleModuleImportStar bug
- Fix #1318: PathUtils.IsNormalizedPath and PathUtils.NormalizePath work incorrectly with UNC
- Fix #1320: Incorrect relative import handling in Python 2x
- Fix #1321: If one user paths is part of the name of another, PathResolverSnapshot is built incorrectly
- Fix paths in tests
- Remove `IModuleManagement.SetUserSearchPaths`

### CODE REVIEW

File: src/Core/Impl/IO/PathUtils.cs [hunk 1]
```diff
@@ -45,7 +45,14 @@ public static bool IsValidFileNameCharacter(char character)
...
         public static bool HasEndSeparator(string path)
-            => !string.IsNullOrEmpty(path) && DirectorySeparators.Contains(path[path.Length - 1]);
+            => !string.IsNullOrEmpty(path) && IsDirectorySeparator(path[path.Length - 1]);
+
+
+        public static bool IsDirectorySeparator(char c) => Array.IndexOf(DirectorySeparators, c) != -1;
+
+        public static bool PathStartsWith(string s, string prefix)
+            => s != null && s.StartsWith(prefix, StringExtensions.PathsStringComparison) &&
+               (s.Length == prefix.Length || IsDirectorySeparator(s[prefix.Length]));
```
Comment (AlexanderSher) [inline src/Core/Impl/IO/PathUtils.cs]: This is fix for #1321, to ensure that part of folder name isn't considered as part of the path.

File: src/Analysis/Ast/Impl/Analyzer/PythonAnalyzerSession.cs [hunk 2]
```diff
@@ -348,7 +348,7 @@ private Task StartAnalysis(IDependencyChainNode<PythonAnalyzerEntry> node, Async
 
         private void LogCompleted(IDependencyChainNode<PythonAnalyzerEntry> node, IPythonModule module, Stopwatch stopWatch, TimeSpan startTime) {
             if (_log != null) {
-                var completed = node != null && module.Analysis is LibraryAnalysis ? "completed" : "completed for library";
+                var completed = node != null && module.Analysis is LibraryAnalysis ? "completed for library" : "completed ";
```
Comment (AlexanderSher) [inline src/Analysis/Ast/Impl/Analyzer/PythonAnalyzerSession.cs]: Correct logging

File: src/Analysis/Ast/Impl/Analyzer/Handlers/ImportHandler.cs [hunk 3]
```diff
@@ -55,25 +55,28 @@ internal sealed partial class ImportHandler : StatementHandler {
...
             // "import fob.oar.baz" is handled as fob = import_module('fob')
-            if (!string.IsNullOrEmpty(asNameExpression?.Name) && importedNamesCount == moduleImportExpression.Names.Count) {
-                Eval.DeclareVariable(asNameExpression.Name, variableModule, VariableSource.Import, asNameExpression);
-            } else if (importedNamesCount > 0 && !string.IsNullOrEmpty(importNames[0]) && _variableModules.TryGetValue(importNames[0], out variableModule)) {
+            if (!string.IsNullOrEmpty(asNameExpression?.Name) && lastModule != default) {
+                Eval.DeclareVariable(asNameExpression.Name, lastModule, VariableSource.Import, asNameExpression);
+            } else if (firstModule != default && !string.IsNullOrEmpty(importNames[0])) {
                 var firstName = moduleImportExpression.Names[0];
-                Eval.DeclareVariable(importNames[0], variableModule, VariableSource.Import, firstName);
+                Eval.DeclareVariable(importNames[0], firstModule, VariableSource.Import, firstName);
```
Comment (AlexanderSher) [inline src/Analysis/Ast/Impl/Analyzer/Handlers/ImportHandler.cs]: This is part of the fix for #1320. Name of the module may not match the name of the variable in Python 2x

File: src/Analysis/Ast/Impl/Analyzer/Handlers/FromImportHandler.cs [hunk 4]
```diff
@@ -82,7 +82,7 @@ internal sealed partial class ImportHandler {
             }
 
             // If __all__ is present, take it, otherwise declare all members from the module that do not begin with an underscore.
-            var memberNames = variableModule.Analysis.StarImportMemberNames ?? variableModule.GetMemberNames().Where(s => !s.StartsWithOrdinal("_"));
+            var memberNames = variableModule.Analysis?.StarImportMemberNames ?? variableModule.GetMemberNames().Where(s => !s.StartsWithOrdinal("_"));
```
Comment (AlexanderSher) [inline src/Analysis/Ast/Impl/Analyzer/Handlers/FromImportHandler.cs]: If `variableModule` represents implicit package, `Analysis` will be `null`.

Comment (MikhailArkhipov) [inline src/Analysis/Ast/Impl/Analyzer/Handlers/FromImportHandler.cs]: Should module have at least empty analysis always? I think there is a lot of code that assumes that.

File: src/Core/Impl/Extensions/StringExtensions.cs [hunk 5]
```diff
@@ -146,9 +146,6 @@ public static string AsQuotedArguments(this IEnumerable<string> args)
             return "\"{0}\"".FormatInvariant(arg);
         }
 
-        public static bool PathStartsWith(this string s, string prefix)
```
Comment (MikhailArkhipov) [inline src/Core/Impl/Extensions/StringExtensions.cs]: Should we have it instead calling `PathUtils.PathStartsWith` ?

File: src/Analysis/Ast/Impl/Analyzer/Handlers/FromImportHandler.cs (hunk 4 above)
Comment (MikhailArkhipov) [reply src/Analysis/Ast/Impl/Analyzer/Handlers/FromImportHandler.cs]: Kinda dangerous... Maybe have it attach empty analysis?

Comment (AlexanderSher) [reply src/Analysis/Ast/Impl/Analyzer/Handlers/FromImportHandler.cs]: Empty analysis requires `IDocument` but there is actually no document - it is an implicit package.

Comment (MikhailArkhipov) [reply src/Analysis/Ast/Impl/Analyzer/Handlers/FromImportHandler.cs]: OK

File: src/Core/Impl/Extensions/StringExtensions.cs (hunk 5 above)
Comment (AlexanderSher) [reply src/Core/Impl/Extensions/StringExtensions.cs]: We have a bit of a mess in our extensions. From my understanding, `StringExtensions` should have things that are applicable to any strings, `PathUtils` should have string operations that are related to paths only, and `FileSystemExtensions` should have things like `FindFile`.

**Pull request merged**